cpp/src/io/utilities/hostdevice_vector.hpp

```diff
@@ -76,7 +83,32 @@ class hostdevice_vector {
     return reinterpret_cast<T const *>(d_data.data()) + offset;
   }
 
+  void host_to_device(cudaStream_t stream)
+  {
+    cudaMemcpyAsync(d_data.data(), h_data, memory_size(), cudaMemcpyHostToDevice, stream);
+    cudaStreamSynchronize(stream);
```

Comment:
I don't think we want to always sync here. There are many places where this is called before another async operation in the same stream. This is perfectly safe (and a lot faster) without a sync here.

Maybe add a parameter that dictates whether or not to sync. Same goes for `device_to_host`, but I'm not sure if defaults should be the same.